src/axom/core/Array.hpp

```diff
@@ -80,68 +113,153 @@ class Array
+   *  with data from an external data buffer, the copy-constructed Array 
+   *  will have a deep copy of the data and own the data copy. 
+   */ 
+  Array( const Array& other, int allocator_id = DEFAULT_ALLOCATOR_ID ); 
+
+  /*! 
+   * \brief Move constructor for an Array instance 
+   *
+   * \note If you use the move constructor on an argument Array with an 
+   *  external data buffer, the move-constructed Array will wrap the external 
+   *  data buffer and the argument Array will be left in a valid empty state. 
+   */ 
+  Array( Array&& other );
 
 /// @}
 
 /// \name External Storage Array Constructors
 /// @{
 
   /*!
-   * \brief Constructs an Array instance with the given number of tuples from
+   * \brief Constructs an Array instance with the given number of elements from
    *  an external data buffer.
    *
    * \param [in] data the external data this Array will wrap.
-   * \param [in] num_tuples the number of tuples in the Array.
-   * \param [in] num_components the number of values per tuple. If not
-   *  specified defaults to 1.
+   * \param [in] num_elements the number of elements in the Array.
    * \param [in] capacity the capacity of the external buffer.
    *
    * \pre data != nullptr
-   * \pre num_tuples > 0
-   * \pre num_components >= 1
+   * \pre num_elements > 0
    *
-   * \post numComponents() == num_components
    * \post getResizeRatio == 0.0
    *
-   * \note a capacity is specified for the number of tuples to store in the
+   * \note a capacity is specified for the number of elements to store in the
    *  array and does not correspond to the actual bytesize.
-   * \note If no capacity or capacity less than num_tuples is specified then
-   *  it will default to the number of tuples.
+   * \note If no capacity or capacity less than num_elements is specified then
+   *  it will default to the number of elements.
    *
    * \note This constructor wraps the supplied buffer and does not own the data.
    *  Consequently, the Array instance cannot be reallocated.
    */
-  Array( T* data, IndexType num_tuples, IndexType num_components=1,
-         IndexType capacity=0 );
+  Array( T* data, IndexType num_elements, IndexType capacity=0 );
+
+/// @}
+
+/// \name Array copy and move operators 
+/// @{ 
+
+  /*! 
+   * \brief Copy assignment operator for Array 
+   * 
+   * \note If you use the copy assignment operator on an argument Array 
+   *  with data from an external data buffer, the copy-assigned Array 
+   *  will have a deep copy of the data and own the data copy. 
+   *
+   * \note The data will be allocated using the allocator ID of the
+   *  copy-assigned Array, not the argument Array.
+   */ 
+  Array& operator=( const Array& other ) 
+  { 
+
+    if ( this != &other ) 
+    { 
+      m_resize_ratio = other.m_resize_ratio; 
+      m_is_external = false;
+      initialize(other.size(), other.capacity()); 
+      std::memcpy(m_data, other.data(),  
```

Comment:
Should this be `axom::copy()`, which also handles different memory spaces, that way if the users wants to shoot themselves in the foot, that can do it correctly....